### ISSUE TITLE
TCR: way to init db on remote VMs? (blocked by #103)

### ISSUE THREAD

**Issue by ndushay** [sul-dlss/preservation_core_catalog#104]
e.g. cap file that can run the db initializations, etc.

possibly:
- users like "ingest" that have write permissions can add checksums to db, but don't have delete or update permissions
- users like "fixity check" can read but can't create, update or delete

**Comment by tallenaz** [comment]
If we want pcc to connect to the remote tcr db, one approach to investigate:
- configure separate db:migrate tasks for each db
- set up separate stanzas for dbs in database.yml
  - if we do this, it would be simpler to manage changes to database.yml in shared_configs, as otherwise, we have to do more work to change how puppet manages that file.
- configure the rails app to know about the separate db configs
see, for example, http://excid3.com/blog/rails-activerecord-multiple-databases-and-migrations/

There may be a prior question whether this is how we want pcc to be able to access tcr.

**Comment by ndushay** [comment]
@tallenaz @julianmorley  based on a conversation with @vivnwong I have the impression that we now want to separate services and use ReST API calls as much as possible.  This implies we might want a very lightweight Rails API wrapper around TCR to make it easy to code the simple ReST calls (CRU).   There's an API-ish spec for the TCR, correct?

**Comment by vivnwong** [comment]
I thought Lynn is going to provide the API spec for the TCR as we discussed in the last analysis/storyTime meeting. Lynn? Julian?

Vivian

On Sep 29, 2017, at 1:22 PM, Naomi Dushay <notifications@github.com<mailto:notifications@github.com>> wrote:


@tallenaz<https://github.com/tallenaz> @julianmorley<https://github.com/julianmorley> based on a conversation with @vivnwong<https://github.com/vivnwong> I have the impression that we now want to separate services and use ReST API calls as much as possible. This implies we might want a very lightweight Rails API wrapper around TCR to make it easy to code the simple ReST calls (CRU). There's an API-ish spec for the TCR, correct?

**Comment by tallenaz** [comment]
I like the idea of a separation of services here -- hooking one rails app up to more than one database can be done, but ReST seems like the more natural approach.

**Comment by LynnMcRae** [comment]
Here's my take on the REST services requirements here in general.  Encapsulating services as RESTful Web Service calls is a layered requirement to make as much of the SDR functionality/APIs available to clients like Argo or robots, or administrators like Ben.  Services -- those common reusable chunks of code that form the basis of any API -- should always start as classes and concerns, as gems and method calls. How they are put together to accomplish a task is an engineering decision, though we see general precedent in some cases, like keeping Argo lean by putting web service calls behind most of its buttons (we expect to fetch PC info about individual objects into Argo this way), or lightweight robots that exist simply to orchestrate service calls. 

This is what we did and continue to do with DOR Services, where there is a nice clean separate **dor-services-app** which does nothing but wrap and deliver DOR methods calls as Web Services (see https://consul.stanford.edu/display/chimera/REST+mappings+for+dor-services+gem).  We could (should?) build on the existing **sdr-services-app** as the means for delivering SDR-PC functionality to remote callers (ignoring naming issues for the moment) and decide cases on their own engineering merit when we bundle code up together into a program vs selectively make calls out. 

I am generally an advocate of exposing as much functionality as possible via RESTful Web Services. It makes a system transparent, demonstrable, open to inspection and remediation by authorized users. It also tends to make units of functionality clear and interactions stable. But this is tempered both by actual need as well as considerations of permissions and knowledge required to exercise interconnected parts properly.  For the TCR, we want a secure data store where only ingest and archiving can write new entries, which is a robot in one case and "the archiving code" in another. Though I like robots that only talk to services, is it worth it here to make it a web service, esp if the archiving code would go directly to the gem?  For read, we have access to the TCR info as a "nice to have" for in the information available to Argo. That would be a REST call, but we haven't decided if there should just be one call to get all the catalog data as a single JSON document or separate calls with, separate forms of output. 

Story time fodder? This is worth exploring.

**Comment by vivnwong** [comment]
Lynn,


hat would be a REST call, but we haven't decided if there should just be one call to get all the catalog data as a single JSON document or separate calls with, separate forms of output.

I think the team (Tony?) is looking for a concrete proposal on TCR API spec from you (and/or Julian) to move forward.  I scheduled the time after Monday 10:30 standup for you, Julian and myself. Whoever want to participate, can join.

Vivian
On Sep 29, 2017, at 4:51 PM, LynnMcRae <notifications@github.com<mailto:notifications@github.com>> wrote:


Here's my take on the REST services requirements here in general. Encapsulating services as RESTful Web Service calls is a layered requirement to make as much of the SDR functionality/APIs available to clients like Argo or robots, or administrators like Ben. Services -- those common reusable chunks of code that form the basis of any API -- should always start as classes and concerns, as gems and method calls. How they are put together to accomplish a task is an engineering decision, though we see general precedent in some cases, like keeping Argo lean by putting web service calls behind most of its buttons (we expect to fetch PC info about individual objects into Argo this way), or lightweight robots that exist simply to orchestrate service calls.

This is what we did and continue to do with DOR Services, where there is a nice clean separate dor-services-app which does nothing but wrap and deliver DOR methods calls as Web Services (see https://consul.stanford.edu/display/chimera/REST+mappings+for+dor-services+gem). We could (should?) build on the existing sdr-services-app as the means for delivering SDR-PC functionality to remote callers (ignoring naming issues for the moment) and decide cases on their own engineering merit when we bundle code up together into a program vs selectively make calls out.

I am generally an advocate of exposing as much functionality as possible via RESTful Web Services. It makes a system transparent, demonstrable, open to inspection and remediation by authorized users. It also tends to make units of functionality clear and interactions stable. But this is tempered both by actual need as well as considerations of permissions and knowledge required to exercise interconnected parts properly. For the TCR, we want a secure data store where only ingest and archiving can write new entries, which is a robot in one case and "the archiving code" in another. Though I like robots that only talk to services, is it worth it here to make it a web service, esp if the archiving code would go directly to the gem? For read, we have access to the TCR info as a "nice to have" for in the information available to Argo. That would be a REST call, but we haven't decided if there should just be one call to get all the catalog data as a single JSON document or separate calls with, separate forms of output.

Story time fodder? This is worth exploring.
